### PR TITLE
rbd: check for cookie support from kernel

### PR DESCRIPTION
**Pull request by pkalever**
# Describe what this PR does #

Currently we only check if the rbd-nbd tool supports cookie feature.
This change will also defend cookie addition based on kernel version

 Signed-off-by: Prasanna Kumar Kalever \<prasanna.kalever@redhat.com\>

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

### CODE REVIEW

**Comment by pkalever** [comment]
@ceph/ceph-csi-contributors this PR passed CI with Quincy ceph release. Dropping the Commit as https://github.com/ceph/ceph-csi/pull/2944 already take cares of it.

File: internal/rbd/rbd_attach.go [hunk 1]
```diff
@@ -71,6 +71,24 @@ const (
 var (
 	hasNBD              = true
 	hasNBDCookieSupport = false
+
+	kernelCookieSupport = []util.KernelVersion{
+		{
+			Version:      5,
+			PatchLevel:   14,
+			SubLevel:     0,
+			ExtraVersion: 0, Distribution: "",
```
Comment (Madhu-1) [inline internal/rbd/rbd_attach.go]: move Distribution to new line

File: internal/rbd/rbd_attach.go [hunk 2]
```diff
@@ -207,21 +225,36 @@ func setRbdNbdToolFeatures() {
...
-			log.WarningLogMsg("rbd-nbd: nbd modprobe failed (%v): %q", err, stderr)
+			log.WarningLogMsg("nbd modprobe failed (%v): %q", err, stderr)
 		}
 	}
-
-	stdout, stderr, err := util.ExecCommand(context.TODO(), rbdTonbd, "--help")
-	if err != nil || stderr != "" {
-		hasNBD = false
-		log.WarningLogMsg("running rbd-nbd --help failed with error:%v, stderr:%s", err, stderr)
+	if hasNBD {
```
Comment (Madhu-1) [inline internal/rbd/rbd_attach.go]: return if the nbd module is not found. we don't need to check cookie support if nbd itself is not present.

File: internal/rbd/rbd_attach.go [hunk 3]
```diff
@@ -207,21 +225,36 @@ func setRbdNbdToolFeatures() {
...
+		log.DefaultLog("nbd module loaded")
 	}
 
-	if strings.Contains(stdout, "--cookie") {
-		hasNBDCookieSupport = true
+	// fetch the current running kernel info
+	release, err := util.GetKernelVersion()
+	if err != nil {
+		log.WarningLogMsg("fetching current kernel version failed (%v)", err)
```
Comment (Madhu-1) [inline internal/rbd/rbd_attach.go:239]: return from here as we dont need to check cookie support as getting kernel version failed.

File: internal/rbd/rbd_attach.go [hunk 4]
```diff
@@ -207,21 +225,36 @@ func setRbdNbdToolFeatures() {
...
 	}
+	if util.CheckKernelSupport(release, kernelCookieSupport) {
+		log.DefaultLog("kernel version %q supports cookie feature", release)
+		// check if the rbd-nbd tool supports cookie
+		stdout, stderr, err := util.ExecCommand(context.TODO(), rbdTonbd, "--help")
+		if err != nil || stderr != "" {
+			hasNBD = false
+			log.WarningLogMsg("running rbd-nbd --help failed with error:%v, stderr:%s", err, stderr)
```
Comment (Madhu-1) [inline internal/rbd/rbd_attach.go]: same here also we can return?

File: internal/rbd/rbd_attach.go [hunk 5]
```diff
@@ -207,21 +225,36 @@ func setRbdNbdToolFeatures() {
...
+		}
 
-	log.DefaultLog("NBD module loaded: %t, rbd-nbd supported features, cookie: %t", hasNBD, hasNBDCookieSupport)
+		if strings.Contains(stdout, "--cookie") {
+			hasNBDCookieSupport = true
```
Comment (Madhu-1) [inline internal/rbd/rbd_attach.go]: we can avoid multiple if else if we return first, suggestion

```go
// set features available with rbd-nbd, and NBD module loaded status.
func setRbdNbdToolFeatures() {
	var stderr string
	// check if the module is loaded or compiled in
	_, err := os.Stat(fmt.Sprintf("/sys/module/%s", moduleNbd))
	if os.IsNotExist(err) {
		// try to load the module
		_, stderr, err = util.ExecCommand(context.TODO(), "modprobe", moduleNbd)
		if err != nil {
			hasNBD = false
			log.WarningLogMsg("nbd modprobe failed (%v): %q", err, stderr)
			return
		}
	}
	log.DefaultLog("nbd module loaded")

	// fetch the current running kernel info
	release, err := util.GetKernelVersion()
	if err != nil {
		log.WarningLogMsg("fetching current kernel version failed (%v)", err)
		return
	}
	if !util.CheckKernelSupport(release, kernelCookieSupport) {
		log.WarningLogMsg("kernel version %q doesn't support cookie feature", release)
		return
	}
	log.DefaultLog("kernel version %q supports cookie feature", release)
	// check if the rbd-nbd tool supports cookie
	stdout, stderr, err := util.ExecCommand(context.TODO(), rbdTonbd, "--help")
	if err != nil || stderr != "" {
		hasNBD = false
		log.WarningLogMsg("running rbd-nbd --help failed with error:%v, stderr:%s", err, stderr)
		return
	}

	if !strings.Contains(stdout, "--cookie") {
		log.WarningLogMsg("rbd-nbd tool doesn't support cookie feature")
		return
	}
	hasNBDCookieSupport = true
	log.DefaultLog("rbd-nbd tool supports cookie feature")

}
```

**Comment by humblec** [comment]
@pkalever thanks for the PR :+1: . If you can recheck on the review comments from @Madhu-1 thats appreciated!

File: internal/rbd/rbd_attach.go (hunk 1 above)
Comment (nixpanic) [reply internal/rbd/rbd_attach.go]: or leave it out completely, maybe only list `Version`, `PatchLevel` and `SubLevel` as three digits for kernel versions are most common

File: internal/rbd/rbd_attach.go (hunk 2 above)
Comment (nixpanic) [reply internal/rbd/rbd_attach.go]: yes, I agree, in general we take the approach of error-checking/validation, and immediately returning if something is not right, only continue if all is OK.

**Review by nixpanic — CHANGES_REQUESTED** [review]

**Review by Madhu-1 — APPROVED** [review]

**Comment by humblec** [comment]
lgtm.. however will wait for Niels.

@nixpanic can you please revisit the PR as there are changes requested..

**Review by nixpanic — APPROVED** [review]

**Pull request merged**